Use watch-based updates for Kubernetes scheduler

DaoWen opened the pull request:
## Changes proposed in this PR

For the Kuberentes scheduler, switch from polling-based scheduler updates to push-based updates on ReplicaSets and Pods via "watches" (K8s' event subscription mechanism).

## Why are we making these changes?

Using watches is more efficient than polling on Kubernetes. This should get us the updates faster, and help avoid performance issues with bottlenecks on the central API server.

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -671,6 +672,107 @@
+           version-key version)
+    version))
+
+(defn start-k8s-watch!
```
This function should be unit tested.

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -671,6 +672,107 @@
+                      "MODIFIED" (swap! watch-state assoc-in [:service-id->pods service-id id] pod)
+                      "DELETED" (swap! watch-state utils/dissoc-in [:service-id->pods service-id id]))))}))
+
+(defn global-rs-state-query
```
This function should be unit tested.

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -671,6 +672,107 @@
+    {:items items
+     :version resource-version}))
+
+(defn global-pods-state-query
```
This function should be unit tested.

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -550,8 +550,9 @@
-  (state [_]
-    {:service-id->failed-instances @service-id->failed-instances-transient-store
+  (state [{:keys [watch-state]}]
+    {:watch-state @watch-state
```
We should have an integration tests that verifies a sufficiently narrow expected modification in the watch state.

DaoWen commented:
@shamsimam - Added test `waiter.scheduler.kubernetes-test/test-start-k8s-watch!`, and fixed the refactoring issue that was causing Travis failures. Ready for another pass, and possibly merging!

DaoWen commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -863,7 +864,9 @@
+          pod-watch-thread (start-pods-watch! scheduler)
+          rs-watch-thread (start-replicasets-watch! scheduler)]
+      (reset! daemon-state {:pod-watch-cancel #(.stop pod-watch-thread)
+                            :rs-watch-cancel #(.stop rs-watch-thread)})
```
I'm aware of this, and there's a reason I chose to use `.stop` despite the deprecation. Unlike the core.async threads we have elsewhere, these threads use non-interruptible blocking i/o and have no guarantee to ever see the cancellation signal that you're suggesting.

DaoWen replied:
To clarify further, *cheshire* is using non-interruptible blocking i/o. Unless we completely re-write the streaming JSON parsing routines, an unclean `Thread/stop` is the only way to kill these.

However, since we never kill the threads outside of the unit tests (we never actually call the functions created here—the unit tests call `start-k8s-watch!` directly), I we could simply drop this entire `daemon-state` atom. If we want something like this again later, we can always add it back in.

I'll just drop the references to `Thread/stop` in the main code for now. We'll still be using `.stop` in the unit tests, but at least we won't have it in the main code, so it won't look as hacky.

shamsimam replied:
I'm happy to see no `Thread/stop` in the main code. I'm aware that cheshire is non-interruptible. The proposed cooperative strategy with the atom is to perform the termination when control comes back to the function when cheshire has returned a single object.

DaoWen replied:
The problem is that the read can block indefinitely (e.g., it might be an hour before the watch thread sees an update), so we can't depend on cooperative termination.

DaoWen commented on waiter/integration/waiter/kubernetes_scheduler_integration_test.clj:
```diff
@@ -0,0 +1,36 @@
+                rs-watch-version' (get-in watch-state-json ["rs-metadata" "version" "watch"])]
+            (is (or (nil? initial-pods-watch-version)
+                    (< initial-pods-snapshot-version initial-pods-watch-version)))
+            (is (<= initial-pods-snapshot-version pods-snapshot-version'))
```
It's `nil` before the initial update, but would never be `nil` here unless all of the previous updates failed. The `:version :watch` values get clobbered every time we do a global-state-of-the-world query, which is why those may be `nil`. I wouldn't expect the watch connection to break in the middle of this test, but I also don't want it to start flaking on us if it does.

DaoWen commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -671,6 +681,120 @@
+                  (update-fn json-object))))
+            (catch Exception e
+              (log/error e "error in" resource-key "state watch thread"))))
+        (log/info "exiting Kubernetes watcher thread" resource-name)))
```
Oops! I should have removed this during my last refactoring. This is now unreachable due to the `(while true ...)`. I'll remove this and push again.

shamsimam replied:
The `try-catch` catches only exceptions, should we have a `System/exit` for other error scenarios?

DaoWen replied:
Good point! I'll add a try–finally around the while loop to handle that case.

DaoWen commented:
@shamsimam - I added the `System/exit` you requested in a `finally` clause after the watcher daemon's infinite loop. I added an `exit-on-error?` option to the `start-k8s-watch!` function so that I can avoid killing the whole JVM when cleaning up daemon threads in the unit tests.

I also refactored how the `api-request-fn` and `streaming-api-request-fn` arguments are passed into `start-k8s-watch!`, so that they're passed in the same map as `exit-on-error?`. This seems much cleaner than what I was previously doing to rebind those functions in the unit tests.
